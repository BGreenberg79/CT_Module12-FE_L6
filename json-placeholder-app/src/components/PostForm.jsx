import React, { useState } from 'react'
import { Form, Button, Container } from 'react-bootstrap'
import axios from 'axios'

const PostForm = () => {

    const [post, setPost] = useState({
        title: '',
        body: '',
        userId: 0
    })

    const addPostAPI = async (post) => {
        const response = await axios.post('https://jsonplaceholder.typicode.com/posts',
            {title: post.title,
            body: post.body,
            userId: post.userId}, 
            {headers: {'Content-type': 'application/json; charset=UTF-8',}}
        )
        return response.data
    }


    const handleSubmit = (event) => {
        event.preventDefault();
        addPostAPI(post)
    }

    return (
        <Container>
            <h1>Add Post</h1>
            <Form onSubmit={handleSubmit}>
                <Form.Group className="mb-3" controlId="title">
                    <Form.Label>Title:</Form.Label>
                    <Form.Control
                        type="text"
                        placeholder="Enter title"
                        onChange={(event) =>
                            setPost({ ...post, title: event.target.value })
                        }
                        value={post.title}
                    />
                </Form.Group>

                <Form.Group className="mb-3" controlId="body">
                    <Form.Label>Body:</Form.Label>
                    <Form.Control
                        type="text"
                        placeholder="Enter body"
                        onChange={(event) =>
                            setPost({ ...post, body: event.target.value })
                        }
                        value={post.body}
                    />
                </Form.Group>

                <Form.Group className="mb-3" controlId="userId">
                    <Form.Label>User Id:</Form.Label>
                    <Form.Control
                        type="number"
                        placeholder="Enter User Id"
                        onChange={(event) =>
                            setPost({ ...post, userId: Number(event.target.value) })
                        }
                        value={post.userId}
                    />
                </Form.Group>


                <Button variant="success" type="submit">
                    Add Post
                </Button>
            </Form>
        </Container>
    )
}

export default PostForm