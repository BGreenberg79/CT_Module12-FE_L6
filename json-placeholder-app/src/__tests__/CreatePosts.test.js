import { TextEncoder, TextDecoder } from 'text-encoding';
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;


import React from 'react';
import '@testing-library/jest-dom';
import PostForm from '../components/PostForm';
import axios from 'axios';
import { render, waitFor, screen, fireEvent } from '@testing-library/react';


jest.mock('axios');

describe('PostForm component', () => {
    test('takes user inputs and then submits to JSON Placeholder API', async () => {
        const mockPost = { title: 'Test Title', body: 'Test Body', userId: 1 }
        axios.post.mockResolvedValue(mockPost);

        render(<PostForm />);

        // Tests user inputs
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/Title:/i), { target: { value: "Test Title" } });
            fireEvent.change(screen.getByLabelText(/Body:/i), { target: { value: "Test Body" } });
            fireEvent.change(screen.getByLabelText(/User Id:/i), { target: { value: "1" } });
        })

        // Simulates Submit button
        await act(async () => {
            fireEvent.click(screen.getByRole('button', { name: /Add Post/i }));
        })

        await waitFor(() => {
            expect(axios.post).toHaveBeenCalledWith('https://jsonplaceholder.typicode.com/posts',
                { title: 'Test Title', body: 'Test Body', userId: 1 },
                { headers: { 'Content-type': 'application/json; charset=UTF-8', } }
            )
        });
    });

    test('matches the snapshot', () => {
        const { asFragment } = render(<PostForm />);
        expect(asFragment()).toMatchSnapshot();
    });
});