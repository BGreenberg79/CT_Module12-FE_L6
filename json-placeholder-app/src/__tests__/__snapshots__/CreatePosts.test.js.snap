// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostForm component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Add Post
    </h1>
    <form
      class=""
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="title"
        >
          Title:
        </label>
        <input
          class="form-control"
          id="title"
          placeholder="Enter title"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="body"
        >
          Body:
        </label>
        <input
          class="form-control"
          id="body"
          placeholder="Enter body"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="userId"
        >
          User Id:
        </label>
        <input
          class="form-control"
          id="userId"
          placeholder="Enter User Id"
          type="number"
          value="0"
        />
      </div>
      <button
        class="btn btn-success"
        type="submit"
      >
        Add Post
      </button>
    </form>
  </div>
</DocumentFragment>
`;
